### PR TITLE
Optional OpenMP, update SZ3 to submodule of offical repo

### PR DESCRIPTION
**Pull request by juntyr**
Follow-up from #3:

This PR:
- makes the dependency on OpenMP optional with a new `openmp` feature
- upgrades sz3-sys to use usize instead of size_t
- upgrades SZ3 to the latest version
- fixes the `DimensionedDataBuilder` to allow length one arrays to be encoded (a bit useless but I actually hit that case once)
- switches the SZ3 source to use a submodule that tracks the official SZ3 main branch
  - this should make future updates a lot cleaner
  - BUT, it makes building slightly more complex as the `version.hpp` file can no longer be patched in as easily
  - alternatively, the SZ3 source could be kept in-repo using a git subtree where each upgrade could mention the exact SZ3 source commit to keep the two coupled (this would also allow more Rust-specific patches to go in faster instead of waiting for them to be upstreamed)
- uses cmake to configure the build (to get the correct version.hpp file in)
  - I don't really like this
  - since SZ3 only uses that information in its binaries, we could provide a dummy file or manually maintain a copy instead

I have not yet merged my fixes to SZ3 into the upstream repo, but this PR prepares sz3-rs to make future upgrades easier.

It would also be fantastic if sz3-sys and sz3-rs could be published to crates.io :)

### CODE REVIEW

File: sz3-sys/src/lib.rs [hunk 1]
```diff
@@ -2,4 +2,6 @@
 #![allow(non_camel_case_types)]
 #![allow(non_snake_case)]
 
+extern crate zstd_sys;
```
Comment (rroohhh) [inline sz3-sys/src/lib.rs:5]: Why was this added? It does not seem necessary to me.

File: sz3-sys/wrapper.hpp [hunk 2]
```diff
@@ -1,31 +1,31 @@
 #include "SZ3/api/sz.hpp"
 
-struct SZ_Config {
```
Comment (rroohhh) [inline sz3-sys/wrapper.hpp:3]: Why were defaults added here? 
These always get fully overwritten by the `sz3` wrapper anyways.

File: sz3-sys/build.rs [hunk 3]
```diff
@@ -91,13 +109,28 @@ fn main() -> Result<(), std::env::VarError> {
     if cfg!(feature = "openmp") {
         if let Some(link) = env::var_os("DEP_OPENMP_CARGO_LINK_INSTRUCTIONS") {
             for i in env::split_paths(&link) {
-                if i.as_os_str().len() == 0 {
-                    continue
+                if i.as_os_str().is_empty() {
+                    continue;
```
Comment (rroohhh) [inline sz3-sys/build.rs:113]:
```suggestion
                    continue
```

**Review by rroohhh — CHANGES_REQUESTED** [review]
Thank you for your PR. 
Overall this looks great, I left some small comments. 

I think the reason I did not add `SZ3` as a submodule is exactly what you discovered regarding the `version.hpp`. I just built `SZ3` locally and then copied over the sources to avoid users requiring `cmake` for this crate. I don't really have a strong opinion on this however, so I would say, lets just go with `SZ3` as a submodule and using `cmake` in `build.rs` for now, its easy enough to change again later if this causes problems. 

Ill prepare a crates.io release, when this PR is merged.

File: src/lib.rs [hunk 4]
```diff
@@ -944,12 +945,13 @@ mod tests {
...
                     let config = Config::new($eb)
                         .lossless($lossless)
-                        .openmp($openmp)
                         .error_bound($eb)
                         .compression_algorithm($ca)
                         .interpolation_algorithm($ia)
                         .quantization_bincount($qb)
                         .block_size($block_size);
+                    #[cfg(feature = "openmp")]
+                    let config = config.openmp($openmp);
```
[the remark this answers is not in the document]
Comment (juntyr) [reply src/lib.rs:963]: Done - the OpenMP tests with `$openmp=true` are now only enabled when `feature="openmp"`, the tests with `$openmp=false` are always enabled.

File: sz3-sys/build.rs [hunk 5]
```diff
@@ -91,13 +109,28 @@ fn main() -> Result<(), std::env::VarError> {
...
+fn copy_dir_all(src: impl AsRef<Path>, dst: impl AsRef<Path>) -> io::Result<()> {
+    fs::create_dir_all(&dst)?;
+
+    for entry in fs::read_dir(src)? {
+        let entry = entry?;
+
+        if entry.file_type()?.is_dir() {
+            copy_dir_all(entry.path(), dst.as_ref().join(entry.file_name()))?;
+        } else {
+            fs::copy(entry.path(), dst.as_ref().join(entry.file_name()))?;
```
[the remark this answers is not in the document]
Comment (juntyr) [reply sz3-sys/build.rs]: I update the code to also update the timestamps

File: sz3-sys/src/lib.rs (hunk 1 above)
Comment (juntyr) [reply sz3-sys/src/lib.rs:6]: I've added a comment to explain why:

When Rust sees the sz3-sys crate, it doesn't see any usage of the zstd_sys crate on the Rust side, which - at least in my WASM compilation - caused ZSTD to not be linked in. Explicitly stating that the sz3-sys crate requires zstd_sys (to be linked in, not e.g. just for macros) seems to fix the issue.

File: sz3-sys/wrapper.hpp (hunk 2 above)
Comment (juntyr) [reply sz3-sys/wrapper.hpp:3]: Good point, I've cleaned up the wrapper to not use any defaults.

File: sz3-sys/build.rs (hunk 3 above)
Comment (juntyr) [reply sz3-sys/build.rs:113]: The semicolon is added by rustfmt (without any changes to its config) on the code. I've left it in for now

File: sz3-sys/src/lib.rs (hunk 1 above)
Comment (rroohhh) [reply sz3-sys/src/lib.rs:6]: I see. For normal compilation https://github.com/apertus-open-source-cinema/sz3-rs/blob/40ed81685098a785d940c1ee60818b4026c50940/sz3-sys/build.rs#L90
is supposed to make rust link against zstd. Do you have an idea why this is not working for WASM?

Comment (juntyr) [reply sz3-sys/src/lib.rs]: I checked in my use case again if this is still needed and it seems that the issue has been resolved. I’m thus happy to remove the extern crate for now - if the issue resurfaces, I know what the bug is and can always submit a minimal PR with an up-to-date failure case.

**Comment by rroohhh** [comment]
Thank you for your PR :)

**Pull request merged**